src/tablenavigation.js

```diff
@@ -0,0 +1,480 @@
+	 */
+	_findTextRangeFromSelection( range, selection, isForward ) {
+		if ( isForward ) {
+			const position = selection.getLastPosition();
+			const lastRangePosition = this._getNearestTextPosition( range, 'backward' );
+
+			if ( !lastRangePosition || position.compareWith( lastRangePosition ) != 'before' ) {
+				return null;
+			}
+
+			return new ModelRange( position, lastRangePosition );
+		} else {
+			const position = selection.getFirstPosition();
+			const firstRangePosition = this._getNearestTextPosition( range, 'forward' );
+
+			if ( !firstRangePosition || position.compareWith( firstRangePosition ) != 'after' ) {
+				return null;
+			}
+
+			return new ModelRange( firstRangePosition, position );
+		}
+	}
+
+	/**
+	 * Basing on provided `boundaries` range, finds first/last (depending on `direction`) element
+	 * that can contain `$text` (according to schema).
+	 *
+	 * @param {module:engine/model/range~Range} boundaries The range to find position in.
+	 * @param {'forward'|'backward'} direction Search direction.
+	 * @returns {module:engine/model/position~Position|null} Nearest selection range or `null` if one cannot be found.
+	 */
+	_getNearestTextPosition( boundaries, direction ) {
+		const schema = this.editor.model.schema;
+		const startPosition = direction == 'forward' ? boundaries.start : boundaries.end;
+
+		const treeWalker = new TreeWalker( { direction, boundaries, startPosition } );
+
+		for ( const { nextPosition } of treeWalker ) {
+			if ( schema.checkChild( nextPosition, '$text' ) ) {
+				return nextPosition;
+			}
+		}
+
+		return null;
+	}
+
+	/**
+	 * Checks if the `modelRange` renders to a single line in the DOM.
+	 *
+	 * @private
+	 * @param {module:engine/model/range~Range} modelRange Current table cell content range.
+	 * @param {Boolean} isForward The expected navigation direction.
+	 * @returns {Boolean}
+	 */
+	_isSingleLineRange( modelRange, isForward ) {
+		const editing = this.editor.editing;
+		const domConverter = editing.view.domConverter;
+
+		if ( isForward && !modelRange.start.isAtEnd ) {
+			// Wrapped lines contain exactly the same position at the end of current line
+			// and at the beginning of next line. That position's client rect is at the end
+			// of current line. In case of caret at first position of the last line that 'dual'
+			// position would be detected as it's not the last line.
+			modelRange = new ModelRange( modelRange.start.getShiftedBy( 1 ), modelRange.end );
+		}
+
+		const viewRange = editing.mapper.toViewRange( modelRange );
+		const domRange = domConverter.viewRangeToDom( viewRange );
+		const rects = Rect.getDomRangeRects( domRange );
+
+		let boundaryVerticalPosition;
+
+		for ( let i = 0; i < rects.length; i++ ) {
+			const idx = isForward ? rects.length - i - 1 : i;
+			const rect = rects[ idx ];
+
+			// We need to check if current `rect` is container for following Rects.
+			const nextRect = rects.find( ( rect, i ) => i > idx && rect.width > 0 );
```

Comment:
When does it happen? Some example would be great.

---

BTW: is this code OK in the first place? Shouldn't the `nextRect` be directional? Like when forward, then checking from `idx` onwards and when backward – in the opposite direction?

Anyway, I think this could be improved in terms of performance. For every `rect` of the main `for()` loop `find()`` will check dozens of `rects` before finding the `nextRect`, while it could be a loop that starts from `idx` instead.